File: vrpc/ui/tray.py
from __future__ import annotations

import logging

import pystray
from pystray import Menu, MenuItem

from ..constants import APP_NAME
from ..i18n import t
from . import icons

logger = logging.getLogger(__name__)


class Tray:
    def __init__(self, settings, callbacks: dict) -> None:
        self.settings = settings
        self.cb = callbacks
        self._active = False
        self.icon = pystray.Icon(
            APP_NAME,
            icon=icons.tray_image(False),
            title=APP_NAME,
            menu=self._build_menu(),
        )

    def _txt(self, key):
        return lambda item: t(self.settings.language, key)

    def _build_menu(self) -> Menu:
        return Menu(
            MenuItem(self._txt("tray_show"), lambda: self.cb["show"](), default=True),
            Menu.SEPARATOR,
            MenuItem(self._txt("rpc_enabled"), lambda: self.cb["toggle_rpc"](),
                     checked=lambda i: self.settings.rpc_enabled),
            MenuItem(self._txt("language"), Menu(
                MenuItem("Türkçe", lambda: self.cb["set_language"]("tr"),
                         checked=lambda i: self.settings.language == "tr", radio=True),
                MenuItem("English", lambda: self.cb["set_language"]("en"),
                         checked=lambda i: self.settings.language == "en", radio=True),
            )),
            MenuItem(self._txt("autostart"), lambda: self.cb["toggle_autostart"](),
                     checked=lambda i: self.settings.autostart),
            Menu.SEPARATOR,
            MenuItem(self._txt("tray_quit"), lambda: self.cb["quit"]()),
        )

    def run(self) -> None:
        self.icon.run_detached()

    def stop(self) -> None:
        try:
            self.icon.stop()
        except Exception:
            pass

    def update(self, active: bool, status_text: str) -> None:
        try:
            if active != self._active:
                self._active = active
                self.icon.icon = icons.tray_image(active)
            self.icon.title = f"{APP_NAME} · {status_text}"
        except Exception as e:
            logger.debug("Tray update error: %s", e)

File: electron/tray.ts
import { Menu, Tray, nativeImage } from 'electron'
import type { Language, Settings } from '@shared/types'
import { APP_NAME, assetPath } from './constants'
import { t } from '@shared/i18n'

export interface TrayDeps {
  showWindow: () => void
  getSettings: () => Settings
  applyPatch: (p: Partial<Settings>) => void
  quit: () => void
}

export class AppTray {
  private tray: Tray
  private active = false

  constructor(private deps: TrayDeps) {
    this.tray = new Tray(this.icon(false))
    this.tray.setToolTip(APP_NAME)
    this.tray.on('click', () => this.deps.showWindow())
    this.tray.on('double-click', () => this.deps.showWindow())
    this.rebuildMenu()
  }

  private icon(active: boolean): Electron.NativeImage {
    const img = nativeImage.createFromPath(
      assetPath(active ? 'game_icon.png' : 'game_icon_idle.png')
    )
    return img.isEmpty() ? nativeImage.createEmpty() : img
  }

  private lang(): Language {
    return this.deps.getSettings().language
  }

  rebuildMenu(): void {
    const s = this.deps.getSettings()
    const lang = s.language
    const menu = Menu.buildFromTemplate([
      {
        label: t(lang, 'tray_show'),
        click: () => this.deps.showWindow()
      },
      { type: 'separator' },
      {
        label: t(lang, 'rpc_enabled'),
        type: 'checkbox',
        checked: s.rpcEnabled,
        click: () => this.deps.applyPatch({ rpcEnabled: !s.rpcEnabled })
      },
      {
        label: t(lang, 'language'),
        submenu: [
          {
            label: 'Türkçe',
            type: 'radio',
            checked: lang === 'tr',
            click: () => this.deps.applyPatch({ language: 'tr', languageDetected: true })
          },
          {
            label: 'English',
            type: 'radio',
            checked: lang === 'en',
            click: () => this.deps.applyPatch({ language: 'en', languageDetected: true })
          }
        ]
      },
      {
        label: t(lang, 'autostart'),
        type: 'checkbox',
        checked: s.autostart,
        click: () => this.deps.applyPatch({ autostart: !s.autostart })
      },
      { type: 'separator' },
      {
        label: t(lang, 'tray_quit'),
        click: () => this.deps.quit()
      }
    ])
    this.tray.setContextMenu(menu)
  }

  update(active: boolean, statusText: string): void {
    if (active !== this.active) {
      this.active = active
      this.tray.setImage(this.icon(active))
    }
    this.tray.setToolTip(`${APP_NAME} · ${statusText}`)
  }

  notify(message: string, title?: string): void {
    try {
      this.tray.displayBalloon({ title: title ?? APP_NAME, content: message })
    } catch {
      
    }
  }

  destroy(): void {
    this.tray.destroy()
  }
}
